File: disciple/win32/win32_combobox.ml
(* OCaml-Win32
 * win32_combobox.ml
 * Copyright (c) 2002 by Harry Chomsky
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free
 * Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *)

open Win32_core

(********************************************************************)
(* User Interface Services: Controls: Combo Boxes *)

external dlg_dir_list_combo_box : wnd:hwnd -> path:string -> combo_id:int -> static_id:int -> options:dlg_dir_list_option list -> string = "dlg_dir_list_combo_box"

external dlg_dir_select_combo_box : wnd:hwnd -> combo_id:int -> (string * bool) = "dlg_dir_select_combo_box"

(* Styles *)

type combobox_style_option =
    CBS_AUTOHSCROLL
  | CBS_DISABLENOSCROLL
  | CBS_DROPDOWN
  | CBS_DROPDOWNLIST
  | CBS_HASSTRINGS
  | CBS_LOWERCASE
  | CBS_NOINTEGRALHEIGHT
  | CBS_OEMCONVERT
  | CBS_OWNERDRAWFIXED
  | CBS_OWNERDRAWVARIABLE
  | CBS_SIMPLE
  | CBS_SORT
  | CBS_UPPERCASE

external combobox_style_of_option : combobox_style_option -> style = "combobox_style_of_option"

let combobox_style_of_options = special_style_of_options ~special_func:combobox_style_of_option

(* Messages *)

let cb_addstring = 0x0143
external issue_cb_addstring : sender:message_sender -> str:string -> int = "issue_cb_addstring"
external pack_cb_addstring_data : data:int32 -> message_contents = "pack_cb_addstring_data"
external unpack_cb_addstring_data : message_result -> int = "unpack_cb_addstring_data"
let issue_cb_addstring_data ~sender ~data = unpack_cb_addstring_data (sender ~msg:(pack_cb_addstring_data ~data))

let cb_deletestring = 0x0144
external pack_cb_deletestring : index:int -> message_contents = "pack_cb_deletestring"
external unpack_cb_deletestring : message_result -> int = "unpack_cb_deletestring"
let issue_cb_deletestring ~sender ~index = unpack_cb_deletestring (sender ~msg:(pack_cb_deletestring ~index))

let cb_dir = 0x0145
external issue_cb_dir : sender:message_sender -> filename:string -> options:dlg_dir_list_option list -> int = "issue_cb_dir"

let cb_findstring = 0x014C
external issue_cb_findstring : sender:message_sender -> start:int -> str:string -> int option = "issue_cb_findstring"
external pack_cb_findstring_data : start:int -> data:int32 -> message_contents = "pack_cb_findstring_data"
external unpack_cb_findstring_data : message_result -> int option = "unpack_cb_findstring_data"
let issue_cb_findstring_data ~sender ~start ~data = unpack_cb_findstring_data (sender ~msg:(pack_cb_findstring_data ~start ~data))

let cb_findstringexact = 0x0158
external issue_cb_findstringexact : sender:message_sender -> start:int -> str:string -> int option = "issue_cb_findstringexact"
external pack_cb_findstringexact_data : start:int -> data:int32 -> message_contents = "pack_cb_findstringexact_data"
external unpack_cb_findstringexact_data : message_result -> int option = "unpack_cb_findstringexact_data"
let issue_cb_findstringexact_data ~sender ~start ~data = unpack_cb_findstringexact_data (sender ~msg:(pack_cb_findstringexact_data ~start ~data))

let cb_getcount = 0x0146
let pack_cb_getcount = pack_simple_message ~id:cb_getcount
external unpack_cb_getcount : message_result -> int = "unpack_cb_getcount"
let issue_cb_getcount ~sender = unpack_cb_getcount (sender ~msg:pack_cb_getcount)

let cb_getcursel = 0x0147
let pack_cb_getcursel = pack_simple_message ~id:cb_getcursel
external unpack_cb_getcursel : message_result -> int option = "unpack_cb_getcursel"
let issue_cb_getcursel ~sender = unpack_cb_getcursel (sender ~msg:pack_cb_getcursel)

let cb_getdroppedcontrolrect = 0x0152
external issue_cb_getdroppedcontrolrect : sender:message_sender -> rect = "issue_cb_getdroppedcontrolrect"

let cb_getdroppedstate = 0x0157
let pack_cb_getdroppedstate = pack_simple_message ~id:cb_getdroppedstate
external unpack_cb_getdroppedstate : message_result -> bool = "unpack_cb_getdroppedstate"
let issue_cb_getdroppedstate ~sender = unpack_cb_getdroppedstate (sender ~msg:pack_cb_getdroppedstate)

let cb_getdroppedwidth = 0x015f
let pack_cb_getdroppedwidth = pack_simple_message ~id:cb_getdroppedwidth
external unpack_cb_getdroppedwidth : message_result -> int = "unpack_cb_getdroppedwidth"
let issue_cb_getdroppedwidth ~sender = unpack_cb_getdroppedwidth (sender ~msg:pack_cb_getdroppedwidth)

let cb_geteditsel = 0x0140
external issue_cb_geteditsel : sender:message_sender -> (int * int) = "issue_cb_geteditsel"

let cb_getextendedui = 0x0156
let pack_cb_getextendedui = pack_simple_message ~id:cb_getextendedui
external unpack_cb_getextendedui : message_result -> bool = "unpack_cb_getextendedui"
let issue_cb_getextendedui ~sender = unpack_cb_getextendedui (sender ~msg:pack_cb_getextendedui)

let cb_gethorizontalextent = 0x015d
let pack_cb_gethorizontalextent = pack_simple_message ~id:cb_gethorizontalextent
external unpack_cb_gethorizontalextent : message_result -> int = "unpack_cb_gethorizontalextent"
let issue_cb_gethorizontalextent ~sender = unpack_cb_gethorizontalextent (sender ~msg:pack_cb_gethorizontalextent)

let cb_getitemdata = 0x0150
external pack_cb_getitemdata : index:int -> message_contents = "pack_cb_getitemdata"
external unpack_cb_getitemdata : message_result -> int32 = "unpack_cb_getitemdata"
let issue_cb_getitemdata ~sender ~index = unpack_cb_getitemdata (sender ~msg:(pack_cb_getitemdata ~index))

let cb_getitemheight = 0x0154
external pack_cb_getitemheight : index:int -> message_contents = "pack_cb_getitemheight"
external unpack_cb_getitemheight : message_result -> int = "unpack_cb_getitemheight"
let issue_cb_getitemheight ~sender ~index = unpack_cb_getitemheight (sender ~msg:(pack_cb_getitemheight ~index))

let cb_getlbtext = 0x0148
external issue_cb_getlbtext : sender:message_sender -> index:int -> max_length:int -> string = "issue_cb_getlbtext"

let cb_getlbtextlen = 0x0149
external pack_cb_getlbtextlen : index:int -> message_contents = "pack_cb_getlbtextlen"
external unpack_cb_getlbtextlen : message_result -> int = "unpack_cb_getlbtextlen"
let issue_cb_getlbtextlen ~sender ~index = unpack_cb_getlbtextlen (sender ~msg:(pack_cb_getlbtextlen ~index))

let cb_getlocale = 0x015A
(* ... *)

let cb_gettopindex = 0x015b
let pack_cb_gettopindex = pack_simple_message ~id:cb_gettopindex
external unpack_cb_gettopindex : message_result -> int = "unpack_cb_gettopindex"
let issue_cb_gettopindex ~sender = unpack_cb_gettopindex (sender ~msg:pack_cb_gettopindex)

let cb_initstorage = 0x0161
external pack_cb_initstorage : count:int -> size:int -> message_contents = "pack_cb_initstorage"
external unpack_cb_initstorage : message_result -> int = "unpack_cb_initstorage"
let issue_cb_initstorage ~sender ~count ~size = unpack_cb_initstorage (sender ~msg:(pack_cb_initstorage ~count ~size))

let cb_insertstring = 0x014A
external issue_cb_insertstring : sender:message_sender -> index:int -> str:string -> int = "issue_cb_insertstring"
external pack_cb_insertstring_data : index:int -> data:int32 -> message_contents = "pack_cb_insertstring_data"
external unpack_cb_insertstring_data : message_result -> int = "unpack_cb_insertstring_data"
let issue_cb_insertstring_data ~sender ~index ~data = unpack_cb_insertstring_data (sender ~msg:(pack_cb_insertstring_data ~index ~data))

let cb_limittext = 0x0141
external pack_cb_limittext : limit:int -> message_contents = "pack_cb_limittext"
let unpack_cb_limittext = ignore
let issue_cb_limittext ~sender ~limit = unpack_cb_limittext (sender ~msg:(pack_cb_limittext ~limit))

let cb_resetcontent = 0x014B
let pack_cb_resetcontent = pack_simple_message ~id:cb_resetcontent
let unpack_cb_resetcontent = ignore
let issue_cb_resetcontent ~sender = unpack_cb_resetcontent (sender ~msg:pack_cb_resetcontent)

let cb_selectstring = 0x014D
external issue_cb_selectstring : sender:message_sender -> start:int -> str:string -> int option = "issue_cb_selectstring"
external pack_cb_selectstring_data : start:int -> data:int32 -> message_contents = "pack_cb_selectstring_data"
external unpack_cb_selectstring_data : message_result -> int option = "unpack_cb_selectstring_data"
let issue_cb_selectstring_data ~sender ~start ~data = unpack_cb_selectstring_data (sender ~msg:(pack_cb_selectstring_data ~start ~data))

let cb_setcursel = 0x014E
external pack_cb_setcursel : index:int -> message_contents = "pack_cb_setcursel"
let unpack_cb_setcursel = ignore
let issue_cb_setcursel ~sender ~index = unpack_cb_setcursel (sender ~msg:(pack_cb_setcursel ~index))

let cb_setdroppedwidth = 0x0160
external pack_cb_setdroppedwidth : width:int -> message_contents = "pack_cb_setdroppedwidth"
external unpack_cb_setdroppedwidth : message_result -> unit = "unpack_cb_setdroppedwidth"
let issue_cb_setdroppedwidth ~sender ~width = unpack_cb_setdroppedwidth (sender ~msg:(pack_cb_setdroppedwidth ~width))

let cb_seteditsel = 0x0142
external pack_cb_seteditsel : start_sel:int -> end_sel:int -> message_contents = "pack_cb_seteditsel"
external unpack_cb_seteditsel : message_result -> unit = "unpack_cb_seteditsel"
let issue_cb_seteditsel ~sender ~start_sel ~end_sel = unpack_cb_seteditsel (sender ~msg:(pack_cb_seteditsel ~start_sel ~end_sel))

let cb_setextendedui = 0x0155
external pack_cb_setextendedui : extended:bool -> message_contents = "pack_cb_setextendedui"
external unpack_cb_setextendedui : message_result -> unit = "unpack_cb_setextendedui"
let issue_cb_setextendedui ~sender ~extended = unpack_cb_setextendedui (sender ~msg:(pack_cb_setextendedui ~extended))

let cb_sethorizontalextent = 0x015e
external pack_cb_sethorizontalextent : extent:int -> message_contents = "pack_cb_sethorizontalextent"
let unpack_cb_sethorizontalextent = ignore
let issue_cb_sethorizontalextent ~sender ~extent = unpack_cb_sethorizontalextent (sender ~msg:(pack_cb_sethorizontalextent ~extent))

let cb_setitemdata = 0x0151
external pack_cb_setitemdata : index:int -> data:int32 -> message_contents = "pack_cb_setitemdata"
external unpack_cb_setitemdata : message_result -> unit = "unpack_cb_setitemdata"
let issue_cb_setitemdata ~sender ~index ~data = unpack_cb_setitemdata (sender ~msg:(pack_cb_setitemdata ~index ~data))

let cb_setitemheight = 0x0153
external pack_cb_setitemheight : index:int -> height:int -> message_contents = "pack_cb_setitemheight"
external unpack_cb_setitemheight : message_result -> unit = "unpack_cb_setitemheight"
let issue_cb_setitemheight ~sender ~index ~height = unpack_cb_setitemheight (sender ~msg:(pack_cb_setitemheight ~index ~height))

let cb_setlocale = 0x0159
(* ... *)

let cb_settopindex = 0x015c
external pack_cb_settopindex : index:int -> message_contents = "pack_cb_settopindex"
external unpack_cb_settopindex : message_result -> unit = "unpack_cb_settopindex"
let issue_cb_settopindex ~sender ~index = unpack_cb_settopindex (sender ~msg:(pack_cb_settopindex ~index))

let cb_showdropdown = 0x014F
external pack_cb_showdropdown : show:bool -> message_contents = "pack_cb_showdropdown"
let unpack_cb_showdropdown = ignore
let issue_cb_showdropdown ~sender ~show = unpack_cb_showdropdown (sender ~msg:(pack_cb_showdropdown ~show))

type compareitemstruct = {
    cis_control_type : control_type;
    cis_control_id : int;
    cis_control : hwnd;
    cis_item_1_id : int;
    cis_item_1_data : int32;
    cis_item_2_id : int;
    cis_item_2_data : int32 }
let wm_compareitem = 0x0039
type wm_compareitem_handler =
    wnd:hwnd ->
    msg:message_contents ->
    cis:compareitemstruct ->
    int message_handler_result
external deliver_wm_compareitem :
    handler:wm_compareitem_handler ->
    wnd:hwnd ->
    msg:message_contents ->
    int message_handler_result = "deliver_wm_compareitem"
external return_wm_compareitem : result:int -> message_result = "return_wm_compareitem"
let on_wm_compareitem = on_message ~id:wm_compareitem ~deliver:deliver_wm_compareitem ~return:return_wm_compareitem
let _ = add_explicit_dialog_return ~id:wm_compareitem

type owner_draw_action =
    ODA_DRAWENTIRE
  | ODA_FOCUS
  | ODA_SELECT
type owner_draw_state =
    ODS_CHECKED
  | ODS_COMBOBOXEDIT
  | ODS_DEFAULT
  | ODS_DISABLED
  | ODS_FOCUS
  | ODS_GRAYED
  | ODS_SELECTED
type drawitemstruct = {
    dis_control_type : control_type;
    dis_control_id : int;
    dis_item_id : int;
    dis_item_action : owner_draw_action list;
    dis_item_state : owner_draw_state list;
    dis_control : hwnd;
    dis_hdc : hdc;
    dis_item_rect : rect;
    dis_item_data : int32 }
let wm_drawitem = 0x002B
type wm_drawitem_handler =
    wnd:hwnd ->
    msg:message_contents ->
    dis:drawitemstruct ->
    unit message_handler_result
external deliver_wm_drawitem :
    handler:wm_drawitem_handler ->
    wnd:hwnd ->
    msg:message_contents ->
    unit message_handler_result = "deliver_wm_drawitem"
external return_wm_drawitem : result:unit -> message_result = "return_wm_drawitem"
let on_wm_drawitem = on_message ~id:wm_drawitem ~deliver:deliver_wm_drawitem ~return:return_wm_drawitem

type measureitemstruct = {
    mis_control_type : control_type;
    mis_control_id : int;
    mis_item_id : int;
    mis_item_data : int32 }
let wm_measureitem = 0x002C
type wm_measureitem_handler =
    wnd:hwnd ->
    msg:message_contents ->
    mis:measureitemstruct ->
    size message_handler_result
external on_wm_measureitem :
    wm_measureitem_handler ->
    wnd:hwnd ->
    msg:message_contents ->
    message_result option = "on_wm_measureitem"

(* Notifications *)

let cbn_closeup = 8
let cbn_dblclk = 2
let cbn_dropdown = 7
let cbn_editchange = 5
let cbn_editupdate = 6
let cbn_killfocus = 4
let cbn_selchange = 1
let cbn_selendcancel = 10
let cbn_selendok = 9
let cbn_setfocus = 3

File: disciple/win32/win32_combobox.c
/* OCaml-Win32
 * win32_combobox.c
 * Copyright (c) 2002 by Harry Chomsky
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free
 * Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include "win32.h"

/********************************************************************/
/* User Interface Services: Controls: Combo Boxes */

CAMLprim value dlg_dir_list_combo_box(value wnd, value path, value combo_id, value static_id, value options)
{
    CAMLparam5(wnd, path, combo_id, static_id, options);
    HWND hwnd;
    char path_buf[_MAX_PATH+1];
    DWORD opt;
    int result;
    CAMLlocal1(val);

    hwnd = Handle_val(wnd);
    strncpy(path_buf, String_val(path), _MAX_PATH);
    path_buf[_MAX_PATH] = '\0';
    opt = flags_val(options, dlg_dir_list_options);
    enter_blocking_section();
    result = DlgDirListComboBox(hwnd, path_buf, Int_val(combo_id), Int_val(static_id), opt);
    leave_blocking_section();
    if (result == 0)
        raise_last_error();
    val = copy_string(path_buf);
    CAMLreturn(val);
}

CAMLprim value dlg_dir_select_combo_box(value wnd, value combo_id)
{
    CAMLparam2(wnd, combo_id);
    HWND hwnd;
    char path_buf[_MAX_PATH+1];
    BOOL result;
    CAMLlocal2(val, v);

    hwnd = Handle_val(wnd);
    path_buf[0] = '\0';
    enter_blocking_section();
    result = DlgDirSelectComboBoxEx(hwnd, path_buf, sizeof(path_buf), Int_val(combo_id));
    leave_blocking_section();
    val = alloc_tuple(2);
    v = copy_string(path_buf);
    Store_field(val, 0, v);
    Store_field(val, 1, Val_bool(result));
    CAMLreturn(val);
}

/* Styles */

static DWORD combobox_style_options[] = {
    CBS_AUTOHSCROLL,
    CBS_DISABLENOSCROLL,
    CBS_DROPDOWN,
    CBS_DROPDOWNLIST,
    CBS_HASSTRINGS,
    CBS_LOWERCASE,
    CBS_NOINTEGRALHEIGHT,
    CBS_OEMCONVERT,
    CBS_OWNERDRAWFIXED,
    CBS_OWNERDRAWVARIABLE,
    CBS_SIMPLE,
    CBS_SORT,
    CBS_UPPERCASE
};
Define_static_constants(combobox_style_options);

CAMLprim value combobox_style_of_option(value option)
{
    return copy_int64((int64)Constant_val(option, combobox_style_options));
}

/* Messages */

CAMLprim value issue_cb_addstring(value sender, value string)
{
    CAMLparam2(sender, string);
    char* s;
    CAMLlocal2(contents, result);

    s = new_string(string);
    contents = pack_message(CB_ADDSTRING, 0, (LPARAM)s);
    result = callback(sender, contents);
    free_string(s);
    CAMLreturn(Val_int(check_combo_box_result(Int32_val(result))));
}

CAMLprim value pack_cb_addstring_data(value data)
{
    return pack_message(CB_ADDSTRING, 0, Int32_val(data));
}

CAMLprim value unpack_cb_addstring_data(value result)
{
    return Val_int(check_combo_box_result(Int32_val(result)));
}

CAMLprim value pack_cb_deletestring(value index)
{
    return pack_message(CB_DELETESTRING, Int_val(index), 0);
}

CAMLprim value unpack_cb_deletestring(value result)
{
    return Val_int(check_combo_box_result(Int32_val(result)));
}

CAMLprim value issue_cb_dir(value sender, value path, value options)
{
    CAMLparam3(sender, path, options);
    char* pth;
    CAMLlocal2(contents, result);

    pth = new_string(path);
    contents = pack_message(CB_DIR, flags_val(options, dlg_dir_list_options), (LPARAM)pth);
    result = callback(sender, contents);
    free_string(pth);
    CAMLreturn(Val_int(check_combo_box_result(Int32_val(result))));
}

CAMLprim value issue_cb_findstring(value sender, value start, value string)
{
    CAMLparam3(sender, start, string);
    char* s;
    CAMLlocal3(contents, result, val);
    int res;

    s = new_string(string);
    contents = pack_message(CB_FINDSTRING, Int_val(start), (LPARAM)s);
    result = callback(sender, contents);
    free_string(s);
    res = Int32_val(result);
    if (res == CB_ERR)
        val = Val_none;
    else
        val = alloc_some(Val_int(res));
    CAMLreturn(val);
}

CAMLprim value pack_cb_findstring_data(value start, value data)
{
    return pack_message(CB_FINDSTRING, Int_val(start), Int32_val(data));
}

CAMLprim value unpack_cb_findstring_data(value result)
{
    int res;

    res = Int32_val(result);
    if (res == CB_ERR)
        return Val_none;
    else
        return alloc_some(Val_int(res));
}

CAMLprim value issue_cb_findstringexact(value sender, value start, value string)
{
    CAMLparam3(sender, start, string);
    char* s;
    CAMLlocal3(contents, result, val);
    int res;

    s = new_string(string);
    contents = pack_message(CB_FINDSTRINGEXACT, Int_val(start), (LPARAM)s);
    result = callback(sender, contents);
    free_string(s);
    res = Int32_val(result);
    if (res == CB_ERR)
        val = Val_none;
    else
        val = alloc_some(Val_int(res));
    CAMLreturn(val);
}

CAMLprim value pack_cb_findstringexact_data(value start, value data)
{
    return pack_message(CB_FINDSTRINGEXACT, Int_val(start), Int32_val(data));
}

CAMLprim value unpack_cb_findstringexact_data(value result)
{
    int res;

    res = Int32_val(result);
    if (res == CB_ERR)
        return Val_none;
    else
        return alloc_some(Val_int(res));
}

CAMLprim value unpack_cb_getcount(value result)
{
    return Val_int(check_combo_box_result(Int32_val(result)));
}

CAMLprim value unpack_cb_getcursel(value result)
{
    int res;

    res = Int32_val(result);
    if (res == CB_ERR)
        return Val_none;
    else
        return alloc_some(Val_int(res));
}

CAMLprim value issue_cb_getdroppedcontrolrect(value sender)
{
    CAMLparam1(sender);
    RECT r;
    CAMLlocal3(contents, result, rect);

    contents = pack_message(CB_GETDROPPEDCONTROLRECT, 0, (LPARAM)&r);
    result = callback(sender, contents);
    if (Int32_val(result) == 0)
        raise_last_error();
    rect = alloc_rect(&r);
    CAMLreturn(rect);
}

CAMLprim value unpack_cb_getdroppedstate(value result)
{
    return Val_bool(Int32_val(result));
}

CAMLprim value unpack_cb_getdroppedwidth(value result)
{
    return Val_int(check_combo_box_result(Int32_val(result)));
}

CAMLprim value issue_cb_geteditsel(value sender)
{
    CAMLparam1(sender);
    CAMLlocal2(contents, val);
    DWORD start, end;

    contents = pack_message(CB_GETEDITSEL, (WPARAM)&start, (LPARAM)&end);
    callback(sender, contents);
    val = alloc_tuple(2);
    Store_field(val, 0, Val_int(start));
    Store_field(val, 1, Val_int(end));
    CAMLreturn(val);
}

CAMLprim value unpack_cb_getextendedui(value result)
{
    return Val_bool(Int32_val(result));
}

CAMLprim value unpack_cb_gethorizontalextent(value result)
{
    return Val_int(Int32_val(result));
}

CAMLprim value pack_cb_getitemdata(value index)
{
    return pack_message(CB_GETITEMDATA, Int_val(index), 0);
}

CAMLprim value unpack_cb_getitemdata(value result)
{
    return copy_int32(check_combo_box_result(Int32_val(result)));
}

CAMLprim value pack_cb_getitemheight(value index)
{
    return pack_message(CB_GETITEMHEIGHT, Int_val(index), 0);
}

CAMLprim value unpack_cb_getitemheight(value result)
{
    return Val_int(check_combo_box_result(Int32_val(result)));
}

CAMLprim value issue_cb_getlbtext(value sender, value index, value max_length)
{
    CAMLparam3(sender, index, max_length);
    char* s;
    CAMLlocal3(contents, result, string);
    int res;

    s = malloc(Int_val(max_length) + 1);
    contents = pack_message(CB_GETLBTEXT, Int_val(index), (LPARAM)s);
    result = callback(sender, contents);
    res = Int32_val(result);
    if (res < 0) {
        free(s);
        check_combo_box_result(res);
    }
    string = copy_string(s);
    free(s);
    CAMLreturn(string);
}

CAMLprim value pack_cb_getlbtextlen(value index)
{
    return pack_message(CB_GETLBTEXTLEN, Int_val(index), 0);
}

CAMLprim value unpack_cb_getlbtextlen(value result)
{
    return Val_int(check_combo_box_result(Int32_val(result)));
}

CAMLprim value unpack_cb_gettopindex(value result)
{
    return Val_int(check_combo_box_result(Int32_val(result)));
}

CAMLprim value pack_cb_initstorage(value count, value size)
{
    return pack_message(CB_INITSTORAGE, Int_val(count), Int_val(size));
}

CAMLprim value unpack_cb_initstorage(value result)
{
    return Val_int(check_combo_box_result(Int32_val(result)));
}

CAMLprim value issue_cb_insertstring(value sender, value index, value string)
{
    CAMLparam3(sender, index, string);
    char* s;
    CAMLlocal2(contents, result);

    s = new_string(string);
    contents = pack_message(CB_INSERTSTRING, Int_val(index), (LPARAM)s);
    result = callback(sender, contents);
    free_string(s);
    CAMLreturn(Val_int(check_combo_box_result(Int32_val(result))));
}

CAMLprim value pack_cb_insertstring_data(value index, value data)
{
    return pack_message(CB_INSERTSTRING, Int_val(index), Int32_val(data));
}

CAMLprim value unpack_cb_insertstring_data(value result)
{
    return Val_int(check_combo_box_result(Int32_val(result)));
}

CAMLprim value pack_cb_limittext(value limit)
{
    return pack_message(CB_LIMITTEXT, Int_val(limit), 0);
}

CAMLprim value issue_cb_selectstring(value sender, value start, value string)
{
    CAMLparam3(sender, start, string);
    char* s;
    CAMLlocal3(contents, result, val);
    int res;

    s = new_string(string);
    contents = pack_message(CB_SELECTSTRING, Int_val(start), (LPARAM)s);
    result = callback(sender, contents);
    free_string(s);
    res = Int32_val(result);
    if (res == CB_ERR)
        val = Val_none;
    else
        val = alloc_some(Val_int(res));
    CAMLreturn(val);
}

CAMLprim value pack_cb_selectstring_data(value start, value data)
{
    return pack_message(CB_SELECTSTRING, Int_val(start), Int32_val(data));
}

CAMLprim value unpack_cb_selectstring_data(value result)
{
    int res;

    res = Int32_val(result);
    if (res == CB_ERR)
        return Val_none;
    else
        return alloc_some(Val_int(res));
}

CAMLprim value pack_cb_setcursel(value index)
{
    return pack_message(CB_SETCURSEL, Int_val(index), 0);
}

CAMLprim value pack_cb_setdroppedwidth(value width)
{
    return pack_message(CB_SETDROPPEDWIDTH, Int_val(width), 0);
}

CAMLprim value unpack_cb_setdroppedwidth(value result)
{
    check_combo_box_result(Int32_val(result));
    return Val_unit;
}

CAMLprim value pack_cb_seteditsel(value start_sel, value end_sel)
{
    return pack_message(CB_SETEDITSEL, 0, MAKELPARAM(Int_val(start_sel), Int_val(end_sel)));
}

CAMLprim value unpack_cb_seteditsel(value result)
{
    check_combo_box_result(Int32_val(result));
    return Val_unit;
}

CAMLprim value pack_cb_setextendedui(value extended)
{
    return pack_message(CB_SETEXTENDEDUI, Bool_val(extended), 0);
}

CAMLprim value unpack_cb_setextendedui(value result)
{
    check_combo_box_result(Int32_val(result));
    return Val_unit;
}

CAMLprim value pack_cb_sethorizontalextent(value extent)
{
    return pack_message(CB_SETHORIZONTALEXTENT, Int_val(extent), 0);
}

CAMLprim value pack_cb_setitemdata(value index, value data)
{
    return pack_message(CB_SETITEMDATA, Int_val(index), Int32_val(data));
}

CAMLprim value unpack_cb_setitemdata(value result)
{
    check_combo_box_result(Int32_val(result));
    return Val_unit;
}

CAMLprim value pack_cb_setitemheight(value index, value height)
{
    return pack_message(CB_SETITEMHEIGHT, Int_val(index), Int_val(height));
}

CAMLprim value unpack_cb_setitemheight(value result)
{
    check_combo_box_result(Int32_val(result));
    return Val_unit;
}

CAMLprim value pack_cb_settopindex(value index)
{
    return pack_message(CB_SETTOPINDEX, Int_val(index), 0);
}

CAMLprim value unpack_cb_settopindex(value result)
{
    check_combo_box_result(Int32_val(result));
    return Val_unit;
}

CAMLprim value pack_cb_showdropdown(value show)
{
    return pack_message(CB_SHOWDROPDOWN, Bool_val(show), 0);
}

CAMLprim value deliver_wm_compareitem(value handler, value wnd, value msg)
{
    CAMLparam3(handler, wnd, msg);
    COMPAREITEMSTRUCT* cis;
    CAMLlocal3(val, v, result);

    cis = (COMPAREITEMSTRUCT*)Lparam(msg);
    val = alloc_tuple(7);
    Store_field(val, 0, Val_constant(cis->CtlType, control_types));
    Store_field(val, 1, Val_int(cis->CtlID));
    v = alloc_handle(cis->hwndItem);
    Store_field(val, 2, v);
    Store_field(val, 3, Val_int(cis->itemID1));
    v = copy_int32(cis->itemData1);
    Store_field(val, 4, v);
    Store_field(val, 5, Val_int(cis->itemID2));
    v = copy_int32(cis->itemData2);
    Store_field(val, 6, v);
    result = callback3(handler, wnd, msg, val);
    CAMLreturn(result);
}

CAMLprim value return_wm_compareitem(value result)
{
    return copy_int32(Int_val(result));
}

static DWORD owner_draw_actions[] = {
    ODA_DRAWENTIRE,
    ODA_FOCUS,
    ODA_SELECT
};
Define_static_constants(owner_draw_actions);

static DWORD owner_draw_states[] = {
    ODS_CHECKED,
    ODS_COMBOBOXEDIT,
    ODS_DEFAULT,
    ODS_DISABLED,
    ODS_FOCUS,
    ODS_GRAYED,
    ODS_SELECTED
};
Define_static_constants(owner_draw_states);

CAMLprim value deliver_wm_drawitem(value handler, value wnd, value msg)
{
    CAMLparam3(handler, wnd, msg);
    DRAWITEMSTRUCT* dis;
    CAMLlocal3(val, v, result);

    dis = (DRAWITEMSTRUCT*)Lparam(msg);
    val = alloc_tuple(9);
    Store_field(val, 0, Val_constant(dis->CtlType, control_types));
    Store_field(val, 1, Val_int(dis->CtlID));
    Store_field(val, 2, Val_int(dis->itemID));
    v = Alloc_flags(dis->itemAction, owner_draw_actions);
    Store_field(val, 3, v);
    v = Alloc_flags(dis->itemState, owner_draw_states);
    Store_field(val, 4, v);
    v = alloc_handle(dis->hwndItem);
    Store_field(val, 5, v);
    v = alloc_handle(dis->hDC);
    Store_field(val, 6, v);
    v = alloc_rect(&dis->rcItem);
    Store_field(val, 7, v);
    v = copy_int32(dis->itemData);
    Store_field(val, 8, v);
    result = callback3(handler, wnd, msg, val);
    CAMLreturn(result);
}

CAMLprim value return_wm_drawitem(value result)
{
    return copy_int32(TRUE);
}

CAMLprim value on_wm_measureitem(value handler, value wnd, value msg)
{
    CAMLparam3(handler, wnd, msg);
    MEASUREITEMSTRUCT* mis;
    CAMLlocal4(val, v, result, ret);

    ret = Val_none;
    if (Int_val(Field(msg, 0)) == WM_MEASUREITEM) {
        mis = (MEASUREITEMSTRUCT*)Lparam(msg);
        val = alloc_tuple(4);
        Store_field(val, 0, Val_constant(mis->CtlType, control_types));
        Store_field(val, 1, Val_int(mis->CtlID));
        Store_field(val, 2, Val_int(mis->itemID));
        v = copy_int32(mis->itemData);
        Store_field(val, 3, v);
        result = callback3(handler, wnd, msg, val);
        if (Is_block(result)) {
            if (Tag_val(result) == 0)
                ret = alloc_some(Field(result, 0));
            else {
                mis->itemWidth = Cx(result);
                mis->itemHeight = Cy(result);
                ret = alloc_some(copy_int32(TRUE));
            }
        }
    }
    CAMLreturn(ret);
}
